### PR TITLE
feat(ui): add extension the top-bar action menu

### PR DESCRIPTION
**Pull request by ashutosh16**
https://github.com/user-attachments/assets/cd2525c2-dbfb-438b-8183-7437d7c87a6c

- Add topbar actionMenuExt.  add custom widgets to perform a specific action

To test
- Checkout custom widget - https://github.com/ashutosh16/argocd-action-menu-extension.git
- yarn install and build and  copy the custom widget in /tmp/extension

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: ui/src/app/applications/components/application-details/application-details.tsx [hunk 1]
```diff
@@ -580,7 +590,18 @@ export class ApplicationDetails extends React.Component<RouteComponentProps<{app
                                                 {title: 'Applications', path: '/applications'},
                                                 {title: <ApplicationsDetailsAppDropdown appName={this.props.match.params.name} />}
                                             ],
-                                            actionMenu: {items: this.getApplicationActionMenu(application, true)},
+                                            actionMenu: {
+                                                items: [...this.getApplicationActionMenu(application, true)]
+                                            },
+                                            actionMenuExtensions: TopBarExtensions.map(ext => (
```
Comment (agilgur5) [inline ui/src/app/applications/components/application-details/application-details.tsx:596]: this doesn't seem to require a new prop -- you can pass an additional `item` in the list above and add a callback (`item.action`) that makes this extension view visible

See also https://github.com/argoproj/argo-ui/pull/576#discussion_r1725980287

Comment (ashutosh16) [reply ui/src/app/applications/components/application-details/application-details.tsx:596]: I would like to make the extension  an opt-in  button. In the UI, additional items always show up as buttons, regardless of whether the extension is needed for that app. If I make change in argo,  I can control the content inside button but not the button, it will always render on the UI.
 For my use case, the button should be displayed when the application has specific  label (prd) and hidden when non-PRD. I don't want to add any logic to determine when to render the extension. I tried the option you mentioned but it doesn't fit the requirement. Add the screenshot for the changes.

Comment (agilgur5) [reply ui/src/app/applications/components/application-details/application-details.tsx:596]: > In the UI, additional items always show up as buttons, regardless of whether the extension is needed for that app. 

The convention for something optional would be to only add it to the items list when it is needed. Adding something that doesn't render would be an anti-pattern

> For my use case, the button should be displayed when the application has specific label (prd) and hidden when non-PRD. I don't want to add any logic to determine when to render the extension.


You have to have that logic somewhere, this just sounds like changing where that logic goes

Comment (ashutosh16) [reply ui/src/app/applications/components/application-details/application-details.tsx:596]: From you [comment](https://github.com/argoproj/argo-cd/pull/19597#issuecomment-2303378924), not sure how I can make extension customizable without adding logic in argocd. I don't want define any template for extension. If Can you provide some  directions ?

Comment (agilgur5) [reply ui/src/app/applications/components/application-details/application-details.tsx:596]: You can have the extension expose an `isVisible` or `shouldDisplay` function (the latter is a little more accurately named). For instance, from the extension registration you have in the docs now:

```diff
- registerTopBarExtension(component: StatusPanelExtensionComponent, title: string, id: string, flyout?: ExtensionComponent, isMiddle: boolean, isNarrow: boolean)
+ registerTopBarExtension(component: StatusPanelExtensionComponent, title: string, id: string, flyout?: ExtensionComponent, isMiddle: boolean, isNarrow: boolean, shouldDisplay?: () => boolean)
```

You can add parameters to the `shouldDisplay` function if that would help as well

Comment (ashutosh16) [reply ui/src/app/applications/components/application-details/application-details.tsx:596]: thank you for the input, changes make sense to me. I make the change and test.

Comment (ashutosh16) [reply ui/src/app/applications/components/application-details/application-details.tsx]: @agilgur5  As you recommended, I don't rely on argo-ui change now, and only argocd has the changes. Can you review the PR?

File: docs/developer-guide/extensions/ui-extensions.md [hunk 2]
```diff
@@ -158,3 +158,64 @@ Below is an example of an extension using the flyout widget:
...
+        title: string,
+        id: string,
+        flyout?: ExtensionComponent,
+        shouldDisplay: (app?: Application) => boolean = () => true,
+        iconClassName?: string,
+        isMiddle = false
+)
+```
+Developer need to pass the function `shouldDisplay` which will be called to determine if the extension should be displayed or not. 
+The function should return true if the extension should be displayed and false otherwise.
```
Comment (agilgur5) [inline docs/developer-guide/extensions/ui-extensions.md:181]:
```suggestion

The callback function `shouldDisplay` should return true if the extension should be displayed and false otherwise:
```

- add a new line after the code block, per markdownlint
- simplify the text here which had several redundant phrases, per [k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-simple-and-direct-language)
- add a colon to segue into the next code block

File: docs/developer-guide/extensions/ui-extensions.md [hunk 3]
```diff
@@ -158,3 +158,64 @@ Below is an example of an extension using the flyout widget:
...
+
+```typescript
+const shouldDisplay =  (app: Application) => {
+  return application.metadata?.labels && application.metadata.labels['application.environmentLabelKey'] === "prd";
+```
+``
```
Comment (agilgur5) [inline docs/developer-guide/extensions/ui-extensions.md:187]:
```suggestion

```

extraneous backticks

File: ui/src/app/applications/components/application-details/application-details.tsx [hunk 4]
```diff
@@ -568,6 +577,7 @@ export class ApplicationDetails extends React.Component<RouteComponentProps<{app
                             });
 
                             const activeExtension = this.state.statusExtensionsMap[this.selectedExtension];
```
Comment (agilgur5) [inline ui/src/app/applications/components/application-details/application-details.tsx:579]: might suggest renaming this to `activeStatusExt` for consistency and less confusion now that there are two types of extensions on this page

File: ui/src/app/applications/components/application-details/application-details.tsx [hunk 5]
```diff
@@ -580,7 +590,14 @@ export class ApplicationDetails extends React.Component<RouteComponentProps<{app
                                                 {title: 'Applications', path: '/applications'},
                                                 {title: <ApplicationsDetailsAppDropdown appName={this.props.match.params.name} />}
                                             ],
-                                            actionMenu: {items: this.getApplicationActionMenu(application, true)},
+                                            actionMenu: {
+                                                items: [
+                                                    ...this.getApplicationActionMenu(application, true),
+                                                    ...(this.state.topBarActionMenuExts
+                                                        ?.filter(ext => ext?.shouldDisplay && ext.shouldDisplay(application))
```
Comment (agilgur5) [inline ui/src/app/applications/components/application-details/application-details.tsx:597]:
```suggestion
                                                        ?.filter(ext => ext.shouldDisplay?.(application))
```
I believe you can further simplify this
Also don't think the optional chaining on `ext?.` is necessary since an item would have to exist for it to be iterated over by `filter`

File: ui/src/app/applications/components/application-details/application-details.tsx [hunk 6]
```diff
@@ -872,6 +889,14 @@ export class ApplicationDetails extends React.Component<RouteComponentProps<{app
                                                 <activeExtension.flyout application={application} tree={tree} />
```
Comment (agilgur5) [inline ui/src/app/applications/components/application-details/application-details.tsx:889]: would also suggest making the above two lines identical to the new ones or vice-versa for consistency.

(could potentially consolidate these two panels as well, since it looks like `isMiddle` is the only different prop right now and it is optional for the panel, but these two extensions could expand with different APIs, so I wouldn't necessarily suggest that just yet)

File: docs/developer-guide/extensions/ui-extensions.md [hunk 7]
```diff
@@ -158,3 +158,64 @@ Below is an example of an extension using the flyout widget:
...
+        iconClassName?: string,
+        isMiddle = false
+)
+```
+Developer need to pass the function `shouldDisplay` which will be called to determine if the extension should be displayed or not. 
+The function should return true if the extension should be displayed and false otherwise.
+
+```typescript
+const shouldDisplay =  (app: Application) => {
+  return application.metadata?.labels && application.metadata.labels['application.environmentLabelKey'] === "prd";
```
Comment (agilgur5) [inline docs/developer-guide/extensions/ui-extensions.md:185]:
```suggestion
  return application.metadata?.labels?.['application.environmentLabelKey'] === "prd";
```

could similarly simplify this with more optional chaining

**Review by agilgur5 — APPROVED** [review]
This makes sense to me 👍 

Just left stylistic comments below.
CD folks should double-check the behavior and docs since I'm not a regular contributor here

File: ui/src/app/applications/components/application-details/application-details.tsx (hunk 1 above)
Comment (agilgur5) [reply ui/src/app/applications/components/application-details/application-details.tsx]: This approach looks a lot better to me 🙂 See my review [below](https://github.com/argoproj/argo-cd/pull/19620#pullrequestreview-2259307947)

File: docs/developer-guide/extensions/ui-extensions.md [hunk 8]
```diff
@@ -158,3 +158,63 @@ Below is an example of an extension using the flyout widget:
...
+)
+```
+
+The callback function `shouldDisplay` should return true if the extension should be displayed and false otherwise:
+
+```typescript
+const shouldDisplay =  (app: Application) => {
+  return application.metadata?.labels?.['application.environmentLabelKey'] === "prd";
+```
+Below is an example of a simple extension with a flyout widget:
```
Comment (agilgur5) [inline docs/developer-guide/extensions/ui-extensions.md:187]:
```suggestion

Below is an example of a simple extension with a flyout widget:
```

missed the new line in my last comment

File: ui/src/app/applications/components/application-details/application-details.tsx [hunk 9]
```diff
@@ -866,10 +883,18 @@ export class ApplicationDetails extends React.Component<RouteComponentProps<{app
                                                 )}
                                         </SlidingPanel>
                                         <SlidingPanel
-                                            isShown={this.selectedExtension !== '' && activeExtension != null && activeExtension.flyout != null}
+                                            isShown={this.selectedExtension !== '' && activeStatusExt != null && activeStatusExt.flyout != null}
+                                            onClose={() => this.setExtensionPanelVisible('')}>
+                                            {this.selectedExtension !== '' && activeStatusExt && activeStatusExt.flyout && (
```
Comment (agilgur5) [inline ui/src/app/applications/components/application-details/application-details.tsx:888]:
```suggestion
                                            {this.selectedExtension !== '' && activeStatusExt?.flyout && (
```

same as `activeTopBarActionMenuExt` below, for consistency

File: ui/src/app/applications/components/application-details/application-details.tsx [hunk 10]
```diff
@@ -866,10 +883,18 @@ export class ApplicationDetails extends React.Component<RouteComponentProps<{app
...
+                                            )}
+                                        </SlidingPanel>
+                                        <SlidingPanel
+                                            isMiddle={activeTopBarActionMenuExt?.isMiddle}
+                                            isShown={this.selectedExtension !== '' && activeTopBarActionMenuExt != null && activeTopBarActionMenuExt.flyout != null}
                                             onClose={() => this.setExtensionPanelVisible('')}>
-                                            {this.selectedExtension !== '' && activeExtension && activeExtension.flyout && (
-                                                <activeExtension.flyout application={application} tree={tree} />
+                                            {this.selectedExtension !== '' && activeTopBarActionMenuExt?.flyout && (
+                                                <activeTopBarActionMenuExt.flyout tree={tree} application={application} />
```
Comment (agilgur5) [inline ui/src/app/applications/components/application-details/application-details.tsx:897]:
```suggestion
                                                <activeTopBarActionMenuExt.flyout application={application} tree={tree} />
```

same ordering as `activeStatusExt` above, for consistency

**Review by agilgur5 — APPROVED** [review]
re-approved with no further stylistic comments. (there is a lint error to fix though)

File: ui/src/app/shared/services/extensions-service.ts [hunk 11]
```diff
@@ -61,6 +62,20 @@ function registerStatusPanelExtension(component: StatusPanelExtensionComponent,
     extensions.eventTarget.emit('statusPanel', ext);
 }
 
+function registerTopBarActionMenuExt(
+    component: TopBarActionMenuExtComponent,
+    title: string,
+    id: string,
+    flyout?: ExtensionComponent,
```
Comment (keithchong) [inline ui/src/app/shared/services/extensions-service.ts:69]: Should this `flyout` be mandatory?  The button click should always result in a flyout sliding panel?

Comment (ashutosh16) [reply ui/src/app/shared/services/extensions-service.ts:69]: yes, good catch.  it should be required, since button ext works  with flyout

File: docs/developer-guide/extensions/ui-extensions.md [hunk 12]
```diff
@@ -158,3 +158,64 @@ Below is an example of an extension using the flyout widget:
...
+The callback function `shouldDisplay` should return true if the extension should be displayed and false otherwise:
+
+```typescript
+const shouldDisplay =  (app: Application) => {
+  return application.metadata?.labels?.['application.environmentLabelKey'] === "prd";
+```
+
+Below is an example of a simple extension with a flyout widget:
+
+```typescript
```
Comment (keithchong) [inline docs/developer-guide/extensions/ui-extensions.md:190]: Since the extensions need to be in javascript, so I think that in this doc, we want users to quickly get these samples up and running. And 'out of the box' (in this case, 'out of the docs'), they should just work.  The Application Status Panel extension sample WITHOUT flyout works fine as is.  I can copy the contents and put them in the extension*.js file in the /tmp/extensions folder, and it works nicely.  However, with the flyout, it doesn't work. This example doesn't work either (registerTopBarExtension is the old name?)

So, I think we could just provide the JS for this example. I got the following to work and there were no errors when the flyout appeared.  Feel free to update the md file with this, and rename the titles/names/etc. if you want.

```
((window) => {
    const shouldDisplay = () => {
      return true;
    };
    const flyout = () => {
      return React.createElement(
              "div",
              { style: { padding: "10px" } },
              "This is a flyout"
      );
    };
    const component = () => {
        return React.createElement(
                "div",
                {
                  onClick: () => flyout()
                },
                "Toolbar Extension Test"
        );
      };
    window.extensionsAPI.registerTopBarActionMenuExt(
            component,
            "Toolbar Extension Test",
            "Toolbar_Extension_Test",
            flyout,
            shouldDisplay,
            '',
            true
    );
  })(window);
```

Also, I got your https://github.com/ashutosh16/argocd-action-menu-extension/tree/main example to work, but I needed to tweak some things there. There's a CORS issue.

Comment (keithchong) [reply docs/developer-guide/extensions/ui-extensions.md:190]: toolbarExtensionTest

Comment (agilgur5) [reply docs/developer-guide/extensions/ui-extensions.md:190]: > Since the extensions need to be in javascript, I think that in this doc, we want users to quickly get these samples up and running.

good point, nice catch! some of the equivalent code blocks are annotated as TS code blocks instead of JS above, so I think it's currently inconsistent in the docs. I'd probably suggest rewriting the "Flyout widget" block in JS. The "Application Status" and "System Level" code blocks are also annotated as TS, but are actually pure JS, so I'd suggest changing the syntax highlight there too for consistency.

In other words, only function type signatures on this page would be in TS, but actual example usage is in JS

Comment (ashutosh16) [reply docs/developer-guide/extensions/ui-extensions.md]: @agilgur5  updated the docs , pls validate

File: docs/developer-guide/extensions/ui-extensions.md [hunk 13]
```diff
@@ -130,26 +130,29 @@ It is also possible to add an optional flyout widget to your extension. It can b
 Below is an example of an extension using the flyout widget:
 
 ```typescript
```
Comment (agilgur5) [inline docs/developer-guide/extensions/ui-extensions.md:132]: This is also an example, so my suggestion was to make this annotation JS and remove the type annotations on the props. I wouldn't change any other formatting or ordering and keep the diff light (right now you split out the code blocks etc, which I wouldn't recommend and actually feels confusing to me)

File: docs/developer-guide/extensions/ui-extensions.md [hunk 14]
```diff
@@ -158,3 +161,65 @@ Below is an example of an extension using the flyout widget:
...
+  };
+  const component = () => {
+    return React.createElement(
+            "div",
+            {
+              onClick: () => flyout()
+            },
+            "Toolbar Extension Test"
+    );
+  };
```
Comment (agilgur5) [inline docs/developer-guide/extensions/ui-extensions.md:205]: I'd move the `flyout` and `component` into the example below as they are examples and pure JS. I would also keep the formatting identical to the prior examples

whereas the `shouldDisplay` part above this can remain in TS as it shows a type signature, same as the registration function

File: docs/developer-guide/extensions/ui-extensions.md [hunk 15]
```diff
@@ -158,3 +161,65 @@ Below is an example of an extension using the flyout widget:
...
+    return true;
+  };
+  window.extensionsAPI.registerTopBarActionMenuExt(
+          component,
+          "Toolbar Extension Test",
+          "Toolbar_Extension_Test",
+          flyout,
+          shouldDisplay,
+          '',
+          true
```
Comment (agilgur5) [inline docs/developer-guide/extensions/ui-extensions.md:222]:
```suggestion
    component,
    "Toolbar Extension Test",
    "Toolbar_Extension_Test",
    flyout,
    shouldDisplay,
    '',
    true
```

keep the indentation consistent with the code above as well as the other examples on this page: 2 space

File: docs/developer-guide/extensions/ui-extensions.md [hunk 16]
```diff
@@ -158,3 +161,65 @@ Below is an example of an extension using the flyout widget:
...
+        isMiddle = false
+)
+```
+
+The callback function `shouldDisplay` should return true if the extension should be displayed and false otherwise:
+
+```typescript
+  const shouldDisplay =  (app: Application) => {
+    return application.metadata?.labels?.['application.environmentLabelKey'] === "prd";
+  };
```
Comment (agilgur5) [inline docs/developer-guide/extensions/ui-extensions.md:189]:
```suggestion
const shouldDisplay = (app: Application) => {
  return application.metadata?.labels?.['application.environmentLabelKey'] === "prd";
};
```

remove some unnecessary spaces

File: docs/developer-guide/extensions/ui-extensions.md [hunk 17]
```diff
@@ -158,3 +161,65 @@ Below is an example of an extension using the flyout widget:
...
+
+```typescript
+registerTopBarActionMenuExt(
+        component: TopBarActionMenuExtComponent,
+        title: string,
+        id: string,
+        flyout?: ExtensionComponent,
+        shouldDisplay: (app?: Application) => boolean = () => true,
+        iconClassName?: string,
+        isMiddle = false
```
Comment (agilgur5) [inline docs/developer-guide/extensions/ui-extensions.md:180]:
```suggestion
  component: TopBarActionMenuExtComponent,
  title: string,
  id: string,
  flyout?: ExtensionComponent,
  shouldDisplay: (app?: Application) => boolean = () => true,
  iconClassName?: string,
  isMiddle = false
```

same as above, consistent 2 space indentation

**Review by agilgur5 — COMMENTED** [review]
Let's keep the formatting and indentation consistent throughout, that makes it easier to read and interpret

**Review by keithchong — APPROVED** [review]

File: docs/developer-guide/extensions/ui-extensions.md [hunk 18]
```diff
@@ -129,32 +129,95 @@ It is also possible to add an optional flyout widget to your extension. It can b
 
 Below is an example of an extension using the flyout widget:
 
-```typescript
+
```
Comment (agilgur5) [inline docs/developer-guide/extensions/ui-extensions.md:132]:
```suggestion
```

unnecessary extra new line added (this would also fail `markdownlint`, which I guess isn't running?)

File: docs/developer-guide/extensions/ui-extensions.md [hunk 19]
```diff
@@ -129,32 +129,95 @@ It is also possible to add an optional flyout widget to your extension. It can b
...
-        style: { padding: "10px" },
-        onClick: () => props.openFlyout()
-      },
-      "Hello World"
+            "div",
+            {
+              style: { padding: "10px" },
+              onClick: () => props.openFlyout()
+            },
+            "Hello World"
```
Comment (agilgur5) [inline docs/developer-guide/extensions/ui-extensions.md:144]:
```suggestion
      "div",
      { 
        style: { padding: "10px" },
        onClick: () => props.openFlyout()
      },
      "Hello World"
```

unnecessary indentation as well, and it's also incorrect and inconsistent -- this adds a 4 space indent in one place to what is otherwise a 2 space code block...

File: docs/developer-guide/extensions/ui-extensions.md [hunk 20]
```diff
@@ -129,32 +129,95 @@ It is also possible to add an optional flyout widget to your extension. It can b
...
     );
   };
   const flyout = () => {
     return React.createElement(
-      "div",
-      { style: { padding: "10px" } },
-      "This is a flyout"
+            "div",
+            { style: { padding: "10px" } },
+            "This is a flyout"
```
Comment (agilgur5) [inline docs/developer-guide/extensions/ui-extensions.md:151]: same as above

File: docs/developer-guide/extensions/ui-extensions.md [hunk 21]
```diff
@@ -129,32 +129,95 @@ It is also possible to add an optional flyout widget to your extension. It can b
...
   };
   window.extensionsAPI.registerStatusPanelExtension(
-    component,
-    "My Extension",
-    "my_extension",
-    flyout
+          component,
+          "My Extension",
+          "my_extension",
+          flyout
```
Comment (agilgur5) [inline docs/developer-guide/extensions/ui-extensions.md:158]: same as above

File: docs/developer-guide/extensions/ui-extensions.md [hunk 22]
```diff
@@ -129,32 +129,95 @@ It is also possible to add an optional flyout widget to your extension. It can b
...
+    );
+  };
+  window.extensionsAPI.registerTopBarActionMenuExt(
+          component,
+          "Toolbar Extension Test",
+          "Toolbar_Extension_Test",
+          flyout,
+          shouldDisplay,
+          '',
+          true
```
Comment (agilgur5) [inline docs/developer-guide/extensions/ui-extensions.md:220]: same as above. this also undoes a previous suggestion I made that you added...

File: docs/developer-guide/extensions/ui-extensions.md [hunk 23]
```diff
@@ -129,32 +129,95 @@ It is also possible to add an optional flyout widget to your extension. It can b
...
+```javascript
+((window) => {
+  const shouldDisplay = () => {
+    return true;
+  };
+  const flyout = () => {
+    return React.createElement(
+            "div",
+            { style: { padding: "10px" } },
+            "This is a flyout"
```
Comment (agilgur5) [inline docs/developer-guide/extensions/ui-extensions.md:201]: same as above

File: docs/developer-guide/extensions/ui-extensions.md [hunk 24]
```diff
@@ -129,32 +129,95 @@ It is also possible to add an optional flyout widget to your extension. It can b
...
+    );
+  };
+  const component = () => {
+    return React.createElement(
+            "div",
+            {
+              onClick: () => flyout()
+            },
+            "Toolbar Extension Test"
```
Comment (agilgur5) [inline docs/developer-guide/extensions/ui-extensions.md:210]: same as above

**Review by agilgur5 — COMMENTED** [review]
There's a bunch of unnecessary indentation issues, some of which I previously corrected for you and were undone.

Please ensure to double-check the diff yourself at least once prior to asking for code review; these issues are fairly easy to spot in the diff and GH points some of them out too. Reviewing your own code is also a best practice.
